File: 2025-09/2025-09-09/uniqueCharacters.js
function allUnique(str) {
  const arr = str.split("");

  return !arr.some((char, index) => arr.indexOf(char) !== index);
}

console.log(allUnique("abc"));
console.log(allUnique("aA"));
console.log(allUnique("QwErTy123!@"));
console.log(allUnique("~!@#$%^&*()_+"));
console.log(allUnique("hello"));
console.log(allUnique("freeCodeCamp"));
console.log(allUnique("!@#*$%^&*()aA"));

// --- TEST SUITE ---

const testsText = `
1. allUnique("abc") should return true.
2. allUnique("aA") should return true.
3. allUnique("QwErTy123!@") should return true.
4. allUnique("~!@#$%^&*()_+") should return true.
5. allUnique("hello") should return false.
6. allUnique("freeCodeCamp") should return false.
7. allUnique("!@#*$%^&*()aA") should return false.
`;

const testsRegex = /(?<number>\d+)\.\s(?<functionCall>.+) should return (?<output>.+?)\.?$/gm;
const testData = [...testsText.matchAll(testsRegex).map(match => match.groups)];

function runTests(testData) {
  console.log("--------------------------");
  console.log("🧪Starting Verification...");
  console.log("--------------------------");

  let allPassed = true;
  let failCount = 0;
  
  testData.forEach(test => {
    const functionCallOutput = eval(test.functionCall);
    const testOutput = eval(test.output);
    let comparison;

    if (Array.isArray(testOutput))
      comparison = arraysEqual(functionCallOutput, testOutput);
    else
      comparison = functionCallOutput === testOutput;

    if (comparison) {
      console.log(`${test.number}.✅PASS - Function Call: ${test.functionCall}`);
    }
    else {
      console.log(`${test.number}.❌FAIL - Function Call: ${test.functionCall}\nExpected: ${testOutput}\nGot: ${functionCallOutput}`);
      allPassed = false;
      failCount ++;
    }
  })

  console.log("----------------------------");

  if (allPassed)
    console.log("🎉SUCCESS: All tests PASSED.");
  else
    console.log(`⚠️WARNING: ${failCount}/${testData.length} tests FAILED.`);
}

function arraysEqual(a, b) {
  if (a.length !== b.length)
    return false;

  for (const index in a) {
    if (Array.isArray(a[index])) {
      if (Array.isArray(b[index])) {
        if (!arraysEqual(a[index], b[index]))
          return false;
      }
      else 
        return false;
    }

    else if (a[index] !== b[index])
      return false;
  }

  return true;
}

runTests(testData);

File: 2025-09/2025-09-09/unique_characters.py
def all_unique(s):
    
    return not any(s.index(char) != index for index, char in enumerate(s))

print(all_unique("abc"))
print(all_unique("aA"))
print(all_unique("QwErTy123!@"))
print(all_unique("~!@#$%^&*()_+"))
print(all_unique("hello"))
print(all_unique("freeCodeCamp"))
print(all_unique("!@#*$%^&*()aA"))

# --- TEST SUITE ---

tests_text = '''
1. all_unique("abc") should return True.
2. all_unique("aA") should return True.
3. all_unique("QwErTy123!@") should return True.
4. all_unique("~!@#$%^&*()_+") should return True.
5. all_unique("hello") should return False.
6. all_unique("freeCodeCamp") should return False.
7. all_unique("!@#*$%^&*()aA") should return False.
'''

import re

tests_regex = re.compile(r"(?P<number>\d+)\.\s(?P<function_call>.+) should return (?P<output>.+?)\.?$", re.MULTILINE)
test_data = [item.groupdict() for item in re.finditer(tests_regex, tests_text)]

def run_tests(test_data):
    print("--------------------------")
    print("🧪 Starting Verification...")
    print("--------------------------")

    all_passed = True
    fail_count = 0

    for test in test_data:
        function_call_output = eval(test['function_call'])
        test_output = eval(test['output'])
        
        if function_call_output == test_output:
            print(f"{test['number']}.✅ PASS - Function Call:\n{test['function_call']}")
        else:
            print(f"{test['number']}.❌ FAIL - Function Call:\n{test['function_call']}\nExpected:\n{test_output}\nGot:\n{function_call_output}")
            all_passed = False
            fail_count += 1

    print("----------------------------");

    if (all_passed):
        print("🎉 SUCCESS: All tests PASSED.");
    else:
        print(f"⚠️ WARNING: {fail_count}/{len(test_data)} tests FAILED.");

run_tests(test_data)